### PR TITLE
installation: CNF Installation (7) Verify new installation process

### PR DESCRIPTION
**Pull request by svteb**
## Description
- New setup spec tests were introduced to verify that the installation method from #2161 works as intended.
- Three distinct issues were discovered and resolved:
  * Helm directories with variable depth levels could not be discovered after initialization (resolved through File.basename).
  * Configs that weren't named cnf-testsuite.yml could not be discovered after initialization (resolved by changing the name internally when copying the file).
  * Existence of manifest folder was not checked (resolved by adding an if check).
- Additionally a common point of failure was added with a failure message that was previously not present.
- The issue #2176 was also discovered during attempted deployment of sample-cnfs/sample-elk-stack which was also added in this commit and is to be used in future spec test.

## For reviewers
**The "validate_config" and "dummy-tag" spec tests are intentionally failing, this is because we are still blocking multiple deployments in one CNF and they are thus considered invalid, these spec tests will start passing again after #2171**. A lot of new sample cnfs were added to test the new installation method making this commit look quite bloated, when reviewing just ignore whatever files that start with `sample-cnfs` (except for the new configs).

## Issues:
Refs: #2168

## How has this been tested:
 - [ ] Covered by existing integration testing
 - [ ] Added integration testing to cover
 - [ ] Verified all A/C passes
     * [ ] develop
     * [ ] master
     * [ ] tag/other branch
 - [ ] Test environment
    * [ ] Shared Packet K8s cluster
    * [ ] New Packet K8s cluster
    * [ ] Kind cluster
 - [ ] Have not tested

## Types of changes:
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
 - [ ] Documentation update

## Checklist:
**Documentation**
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] No updates required.

**Code Review**
- [ ] Does the test handle fatal exceptions, ie. rescue block

**Issue**
- [ ] Tasks in issue are checked off

### CODE REVIEW

File: src/tasks/utils/cnf_installation/config_versions/config_v2.cr [hunk 1]
```diff
@@ -60,9 +60,9 @@ module CNFInstall
         end
         
         # To be removed with new installation process.
-        if @helm_charts.size + @helm_dirs.size + @manifests.size > 1 
-          raise YAML::Error.new("Multiple deployments are not supported yet")
-        end
+        # if @helm_charts.size + @helm_dirs.size + @manifests.size > 1 
+        #  raise YAML::Error.new("Multiple deployments are not supported yet")
+        # end
```
Comment (kosstennbl) [inline src/tasks/utils/cnf_installation/config_versions/config_v2.cr:65]: As multiple deployments aren't ready on regular testcatalog performance, i think that this code should not be commented.
Please uncomment that code and disable (not delete) new spec tests with multiple deployments

File: src/tasks/utils/cnf_installation/install_common.cr [hunk 2]
```diff
@@ -97,8 +97,11 @@ module CNFInstall
     deployment_managers.each do |deployment_manager|
       deployment_name = deployment_manager.deployment_name
 
-      stdout_success "Installing deployment #{deployment_name}"
-      deployment_manager.install()
+      stdout_success "Installing deployment \"#{deployment_name}\"."
+      if !deployment_manager.install()
```
Comment (kosstennbl) [inline src/tasks/utils/cnf_installation/install_common.cr:101]: Maybe storing result of `deployment_manager.install()` in a separate variable before if condition would help with better readability of code.

**Review by kosstennbl — CHANGES_REQUESTED** [review]
Haven't checked the sample CNFs, but i suppose that they should work fine,
Two notes:

**Review by kosstennbl — APPROVED** [review]
LGTM,
Dummy-tag specs are preparation for step 8 and are meant to fail.
Validate_config spec is failing because of addition of multiple-deployment CNF for dummy-tag tests whilst multiple deployments are not supported yet for regular functionality of testsuite. So, CI failures are expected and don't point to any actual problems.

File: sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+{{- $presets := dict 
+  "nano" (dict 
+      "requests" (dict "cpu" "100m" "memory" "128Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "150m" "memory" "192Mi" "ephemeral-storage" "2Gi")
+   )
+  "micro" (dict 
+      "requests" (dict "cpu" "250m" "memory" "256Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "375m" "memory" "384Mi" "ephemeral-storage" "2Gi")
+   )
+  "small" (dict
```
Comment (collivier) [inline sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl:24]:
```suggestion
  "small" (dict
```

File: sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+   )
+  "micro" (dict 
+      "requests" (dict "cpu" "250m" "memory" "256Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "375m" "memory" "384Mi" "ephemeral-storage" "2Gi")
+   )
+  "small" (dict 
+      "requests" (dict "cpu" "500m" "memory" "512Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "750m" "memory" "768Mi" "ephemeral-storage" "2Gi")
+   )
+  "medium" (dict
```
Comment (collivier) [inline sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl:28]:
```suggestion
  "medium" (dict
```

File: sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+   )
+  "small" (dict 
+      "requests" (dict "cpu" "500m" "memory" "512Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "750m" "memory" "768Mi" "ephemeral-storage" "2Gi")
+   )
+  "medium" (dict 
+      "requests" (dict "cpu" "500m" "memory" "1024Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "750m" "memory" "1536Mi" "ephemeral-storage" "2Gi")
+   )
+  "large" (dict
```
Comment (collivier) [inline sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl:32]:
```suggestion
  "large" (dict
```

File: sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+   )
+  "medium" (dict 
+      "requests" (dict "cpu" "500m" "memory" "1024Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "750m" "memory" "1536Mi" "ephemeral-storage" "2Gi")
+   )
+  "large" (dict 
+      "requests" (dict "cpu" "1.0" "memory" "2048Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "1.5" "memory" "3072Mi" "ephemeral-storage" "2Gi")
+   )
+  "xlarge" (dict
```
Comment (collivier) [inline sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl:36]:
```suggestion
  "xlarge" (dict
```

File: sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+   )
+  "large" (dict 
+      "requests" (dict "cpu" "1.0" "memory" "2048Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "1.5" "memory" "3072Mi" "ephemeral-storage" "2Gi")
+   )
+  "xlarge" (dict 
+      "requests" (dict "cpu" "1.0" "memory" "3072Mi" "ephemeral-storage" "50Mi")
+      "limits" (dict "cpu" "3.0" "memory" "6144Mi" "ephemeral-storage" "2Gi")
+   )
+  "2xlarge" (dict
```
Comment (collivier) [inline sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_resources.tpl:40]:
```suggestion
  "2xlarge" (dict
```

File: sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_utils.tpl [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+{{- $value := "" -}}
+{{- $latestObj := $.context.Values -}}
+{{- range $splitKey -}}
+  {{- if not $latestObj -}}
+    {{- printf "please review the entire path of '%s' exists in values" $.key | fail -}}
+  {{- end -}}
+  {{- $value = ( index $latestObj . ) -}}
+  {{- $latestObj = $value -}}
+{{- end -}}
+{{- printf "%v" (default "" $value) -}}
```
Comment (collivier) [inline sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_utils.tpl:49]:
```suggestion
{{- printf "%v" (default "" $value) -}}
```

File: sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_utils.tpl [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+{{- define "common.utils.getKeyFromList" -}}
+{{- $key := first .keys -}}
+{{- $reverseKeys := reverse .keys }}
+{{- range $reverseKeys }}
+  {{- $value := include "common.utils.getValueFromKey" (dict "key" . "context" $.context ) }}
+  {{- if $value -}}
+    {{- $key = . }}
+  {{- end -}}
+{{- end -}}
+{{- printf "%s" $key -}}
```
Comment (collivier) [inline sample-cnfs/sample_multiple_deployments/nginx/charts/common/templates/_utils.tpl:66]:
```suggestion
{{- printf "%s" $key -}}
```

**Review by collivier — APPROVED** [review]

**Pull request merged**